### ISSUE TITLE
Add 3d CAD generation

### ISSUE THREAD

**Issue by DanThePotato** [pewdiepie-archdaemon/odysseus#3625]
### Prerequisites

- [x] I searched [open issues](https://github.com/pewdiepie-archdaemon/odysseus/issues?q=is%3Aissue+is%3Aopen) and this has not already been proposed.
- [x] I searched [discussions](https://github.com/pewdiepie-archdaemon/odysseus/discussions) and this is not already being debated there.
- [x] This is a concrete, actionable proposal — not a vague "it would be nice if..." request.

### Area

Other

### Problem or Motivation

Integrating local 3D CAD generation into Odysseus solves critical bottlenecks in engineering and manufacturing. It moves the technology beyond a novel gimmick by directly addressing four real-world problems and enabling specific, high-value use cases.Real-World Problems SolvedThe Industrial Espionage & Data Privacy BottleneckProblem: Corporate legal teams ban engineers from using cloud-based AI tools (like ChatGPT or Midjourney) because pasting proprietary dimensions or novel inventions leaks intellectual property (IP) to external servers.Solution: Because Odysseus runs locally, aerospace, defense, and medical R&D engineers can safely use AI to generate blueprints for highly confidential parts without risking data leaks, compliance penalties, or corporate espionage.The CAD Software Learning Curve (The "CAD Tax")Problem: Small business owners, inventors, and hobbyists often abandon hardware ideas because professional CAD software (SolidWorks, Fusion360) requires hundreds of hours to master, and hiring a freelance designer costs thousands of dollars.Solution: It lowers the barrier to entry to zero. A user can type what they need in plain English, bypassing complex user interfaces, constraint mapping, and sketch tools entirely.The Uneditable AI Output ProblemProblem: Existing 2D image generators (like Stable Diffusion) can show a picture of a bracket, but you cannot extract dimensions or manufacture it. Existing 3D mesh generators output "point clouds" or messy OBJ files that are impossible to edit, measure, or use in professional engineering.Solution: By generating parametric code (like OpenSCAD or CadQuery) instead of raw meshes, Odysseus provides mathematical files. If the generated screw hole is 4mm and the user needs 5mm, they can simply change one number in the code.The Iteration Lag TimeProblem: Manually testing variations of a single mechanical part—such as changing a bracket's length by 5mm increments to find the perfect fit—takes hours of clicking, dragging, and saving in traditional software.Solution: It automates batch iteration. Engineers can ask the local AI to "generate 5 variations of this casing with thicknesses ranging from 2mm to [1](https://www.pcworld.com/article/2919289/windows-ml-is-ready-to-save-local-ai.html)0mm," compressing days of design exploration into seconds.Specific Use Cases EnabledOn-Demand Replacement Parts for Maintenance (MRO)Scenario: A technician in a remote factory or a cargo ship needs to fix a machine, but a specific plastic gear or mounting bracket has snapped. Shipping the part will take two weeks and halt production.Action: The technician describes the broken part to Odysseus: "A spur gear, [2](https://www.blog.rendair.ai/creating-3d-architectural-visualizations-ai-approach/)0 teeth, 50mm outer diameter, with a 10mm D-profile center bore." Odysseus generates the file locally, and the technician immediately fabricates it using an on-site 3D printer or CNC mill.Generative Fixes for the Maker MovementScenario: A hobbyist wants to mount a specific webcam to a non-standard webcam stand they bought online. No such adapter exists commercially.Action: They prompt Odysseus: "Design a male-to-female adapter. Side A screws into a standard 1/4-inch tripod thread. Side B clips tightly onto a 20mm square tube." They get a flawless, 3D-printable STL file tailored exactly to their unique problem.Instant Enclosures for Electronics EngineeringScenario: An electrical engineer just finished prototyping a custom circuit board (PCB). They need a plastic box to protect it, with precise cutouts for a USB-C port and a power switch.Action: They feed the PCB dimensions into Odysseus: "Create a two-part snap-fit enclosure for a 50mm x 80mm board, including a 4mm x [1](https://www.autodesk.com/products/fusion-360/blog/autodesk-the-first-to-pioneer-cloud-cad-with-over-55-million-users-today/)0mm rectangular cutout on the left face for a port." The housing is generated instantly.

### Proposed Solution

To make this feature concrete, Odysseus will introduce a dual-pane UI workspace and utilize a text-to-code generation loop. Instead of generating raw, uneditable 3D meshes, Odysseus will generate deterministic CadQuery (Python-based) code. This code will compile locally into precise 3D geometry.[1](https://medium.com/@orfizm42/vibe-modeling-a-new-approach-to-designing-3d-models-with-llms-485ed91f2591). UI Sketch (The Dual-Pane Workspace)The chat interface expands into a split-screen workspace when a CAD request is detected.text+---------------------------------------------------------------------------------------+

|  Odysseus AI  [Local Model: DeepSeek-Coder-7B]                                     [X] |
+---------------------------------------------------+-----------------------------------+

| LEFT PANE: Chat & Interactive Terminal             | RIGHT PANE: Local 3D Viewer       |
|                                                   |                                   |
| User: Generate a custom wall mount bracket for a  |  +-----------------------------+  |
| 30mm pipe. It needs two 5mm screw holes spaced    |  |                             |  |
| 60mm apart on the base tab.                       |  |            /---\            |  |
|                                                   |  |           |     |           |  |
| Odysseus: Generating CadQuery code...             |  |         --+     +--         |  |
| Compiling geometry... Success!                    |  |        (o)       (o)        |  |
|                                                   |  |         ---------           |  |
| [▶ Run Code] [🔄 Regenerate]                      |  |                             |  |
|                                                   |  |      [3D WebGL Canvas]      |  |
| Terminal Output:                                  |  +-----------------------------+  |
| > Bounding Box: 80mm x 40mm x 35mm                |  | Rotate | Pan | Zoom | Inspect |
| > Volume: [1](https://support.claude.com/en/articles/12138966-release-notes)4.2 cm³                                |  +-----------------------------+  |
|                                                   |  | [Export STEP]  [Export STL] |  |
+---------------------------------------------------+-----------------------------------+
Use code with caution.2. The API Shape (Backend Integration)Odysseus core will expose a new cad service module. This API handles the execution of the generated Python code within a safe, sandboxed environment on the user's machine, returning the geometric data and metadata to the front end.POST /api/cad/generatePayload Sent to Local LLM Engine:json{
  "prompt": "A wall mount bracket for a 30mm pipe with two 5mm screw holes spaced 60mm apart on the base tab.",
  "engine": "cadquery",
  "unit_system": "metric"
}
Use code with caution.Response Schema From Local Engine:json{
  "success": true,
  "code": "import cadquery as cq\n\n# Parameters\npipe_d = 30.0\nhole_d = 5.0\n...\nresult = cq.Workplane('XY')...",
  "metadata": {
    "dimensions": { "width": 80.0, "length": 40.0, "height": 35.0 },
    "volume_cm3": [1](https://dev.to/aws/dev-track-spotlight-how-amazon-teams-use-ai-assistants-to-accelerate-development-dev403-1bh4)4.25,
    "warnings": []
  },
  "visual_data": {
    "format": "threejs_json",
    "mesh_data": "{...serialized WebGL vertices and faces...}"
  }
}
Use code with caution.3. Under the Hood: Generated Code SnippetWhen the user gives the prompt, Odysseus compiles and executes code natively using the cadquery library. This is the exact code the local LLM generates behind the scenes to build the part deterministically:pythonimport cadquery as cq

# 1. Define Design Parameters (Extracted from User Prompt)
pipe_diameter = 30.0
wall_thickness = 4.0
base_width = 80.0
base_thickness = 5.0
hole_distance = 60.0
hole_diameter = 5.0

# 2. Build the Main Base Tab
base = (
    cq.Workplane("XY")
    .box(base_width, 20.0, base_thickness)
    .faces("<Z")
    .workplane()
    .rect(hole_distance, 0.0)
    .vertices()
    .hole(hole_diameter)
)

# 3. Build the Pipe Clamp Ring
clamp = (
    cq.Workplane("XZ")
    .center(0, base_thickness)
    .circle((pipe_diameter / 2) + wall_thickness)
    .extrude(20.0)
    .faces(">Y")
    .workplane()
    .circle(pipe_diameter / 2)
    .cutThruAll()
)

# 4. Combine Features into a Single Solid Component
final_bracket = base.union(clamp)

# 5. Export to Local Workspace
cq.exporters.export(final_bracket, "output.step")
cq.exporters.export(final_bracket, "output.stl")
Use code with caution.4. Exact Behaviour & User FlowTrigger: The user requests a physical object (e.g., "Make a camera mount..."). Odysseus automatically shifts from standard chat markdown to the CAD Workspace Layout.Generation: The local LLM outputs raw Python CadQuery code.Compilation: Odysseus passes this code to an internal, sandboxed Python environment pre-packaged with cadquery and OCP (Open CASCADE Technology).Rendering: The script outputs a temporary STEP/GLTF file. The UI catches this file and renders it live in the right-hand panel using Three.js.Modification Loop: The user can either modify the code blocks directly in the chat panel or text-prompt a change (e.g., "Make the holes 6mm instead"). The LLM updates only the parameters at the top of the script and reruns it instantly.

### Alternatives Considered

1. Direct STL Mesh Generation (e.g., Shap-E, Point-E, MeshGPT)The Approach: Train or fine-tune a local model to directly output a 3D point cloud or raw polygonal mesh (.stl or .obj file).Why it was ruled out:Uneditable: Raw meshes are just a collection of thousands of triangles. You cannot easily change a bolt hole from 4mm to 5mm because there are no mathematical parameters; it is just a hollow shell.Low Precision: Neural networks that generate meshes directly suffer from "fuzzy geometry." Right angles are wavy, holes are rarely perfectly circular, and flat faces are uneven. This makes them useless for mechanical parts.Resource Intensive: Running local 3D diffusion or autoregressive mesh models requires massive VRAM, slowing down the host machine significantly.[2](https://dienamics.com.au/blog/cad-file-formats-who-makes-them-what-they-do-which-is-best-for-you/). Neural Radiance Fields (NeRFs) / 3D Gaussian SplattingThe Approach: Generate a 3D scene representation from a text prompt or 2D image sequence.Why it was ruled out:Purely Visual: NeRFs and Gaussian Splats are designed for rendering graphics, games, and visual effects. They contain zero geometric, structural, or volumetric data required for manufacturing or engineering. You cannot slice a Gaussian splat for a 3D printer.3. Fine-tuning an LLM on Raw STEP/IGES Text FilesThe Approach: Train an LLM on the raw text-based mathematical logs of exchange files like STEP or IGES.Why it was ruled out:Lack of Tokens/Context: STEP files are incredibly verbose and structurally repetitive. A simple cube with a hole can be thousands of lines of cryptographic-looking coordinate data. An LLM would exhaust its context window on simple geometries and struggle to maintain mathematical logic over long sequences.Why the Proposed CadQuery/OpenSCAD Approach WinsBy generating high-level programming code rather than direct geometry:It leverages existing LLM strengths: LLMs are already world-class at generating, debugging, and refactoring Python code.It handles the heavy lifting off-model: The LLM only writes a 30-line script. The intense math of rendering thousands of 3D vertices is handed off to an established, highly optimized geometry engine (Open CASCADE or OpenSCAD) running locally on the CPU/GPU.Existing Workarounds (How Users Do This Today)Currently, users have to jump through multiple clunky hops to achieve this locally, creating a massive opportunity for Odysseus to streamline the process.The "Prompt-to-Editor" Hop (The Code-Savvy Workaround)Prompting: A user opens Odysseus and prompts the local LLM: "Write me an OpenSCAD script for a 30mm pipe bracket."Copy/Paste: The user manually copies the code block from Odysseus.External Software: They open a completely separate desktop application like OpenSCAD or a Python environment with CadQuery installed.Compile & Loop: They paste the code, render it, notice a mistake, manually tweak the code or go back to Odysseus, paste the error message, and repeat.Why it's broken: It lacks a tight feedback loop. Non-programmers cannot participate because they don't know how to install or debug Python environments or external CAD scripts.The "[2](https://www.youtube.com/watch?v=hPPqodHdIwo)D Image to SVG Vector" Hack (The Slicing Workaround)Generation: A user uses a local 2D image generator (like Stable Diffusion) to create a top-down black-and-white silhouette of a mechanical part.Vectorization: They use an external tool (like inkscape or an online converter) to convert the image into a 2D vector file (.svg).Extrusion: They import that SVG into an engineering tool (like Blender or Fusion360) and manually extrude it into the 3D Z-axis.Why it's broken: It is limited to flat, 2.5D shapes. It cannot handle complex 3D features like internal channels, countersunk holes, or multi-axis geometry.

### Prior Art / Related Issues

https://github.com/pewdiepie-archdaemon/odysseus/discussions/436

### Are you willing to implement this?

Partially — I can help but need guidance
